.github/workflows/test.yml

```diff
@@ -0,0 +1,103 @@
+  DOCKER_IMAGE: ${{ github.repository }}
+  DOCKER_REGISTRY_HOST: ${{ secrets.DOCKER_REGISTRY_HOST }}
+  DOCKER_REGISTRY_USERNAME: ${{ github.repository_owner }}
+  DOCKER_REGISTRY_TOKEN: ${{ secrets.DOCKER_REGISTRY_TOKEN }}
+  RUBY_VERSION: 2.7.1
+  NODE_VERSION: 12
+
+jobs:
+  test:
+    name: Test ${{ matrix.variant }} variant
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        variant: [web, api]
+    steps:
+      - name: Cancel previous runs
+        uses: styfle/cancel-workflow-action@0.5.0
+        with:
+          access_token: ${{ github.token }}
+
+      - uses: actions/checkout@v2
+
+      - uses: nimblehq/branch-tag-action@v1.1
+
+      - uses: ruby/setup-ruby@v1.45.0
+        with:
+          ruby-version: ${{ env.RUBY_VERSION }}
+
+      - uses: actions/setup-node@v2-beta
+        with:
+          node-version: ${{ env.NODE_VERSION }}
+
+      - name: Gems cache
+        uses: actions/cache@v2
+        with:
+          path: /home/runner/.rubies/ruby-${{ env.RUBY_VERSION }}
+          key: ${{ runner.os }}-gems-${{ env.RUBY_VERSION }}-${{ hashFiles('**/Gemfile.lock') }}
+          restore-keys: |
+            ${{ runner.os }}-gems-${{ env.RUBY_VERSION }}-
+
+      - name: Create app folder
+        if: ${{ matrix.variant == 'web' }}
+        run: mkdir $APP_NAME
+
+      - name: Get yarn cache directory path
+        if: ${{ matrix.variant == 'web' }}
+        id: yarn-cache-dir-path
+        run: echo "::set-output name=yarn_cache_dir::$(yarn cache dir)"
+
+      - name: npm packages cache
+        if: ${{ matrix.variant == 'web' }}
+        uses: actions/cache@v2
+        with:
+          path: |
+            ${{ steps.yarn-cache-dir-path.outputs.yarn_cache_dir }}
+            **/yarn.lock
+            **/node_modules
+          key: ${{ runner.os }}-npm-packages-${{ hashFiles('**/package.json*') }}
+          restore-keys: |
+            ${{ runner.os }}-npm-packages-
+
+      - name: Create rails app
+        run: |
+          gem install rails --no-document
+          make create_${{ matrix.variant }}
+
+      - name: Docker login
+        run: echo "$DOCKER_REGISTRY_TOKEN" | docker login $DOCKER_REGISTRY_HOST --username=$DOCKER_REGISTRY_USERNAME --password-stdin
+
+      - name: Docker pull
+        if: env.BRANCH_TAG != 'latest' && env.BRANCH_TAG != 'development'
+        run: |
+          export BRANCH_TAG=${{ env.BRANCH_TAG }}-${{ matrix.variant }}
+          cd $APP_NAME
+          docker-compose pull test || true
+
+      - name: Docker build
+        run: |
+          export BRANCH_TAG=${{ env.BRANCH_TAG }}-${{ matrix.variant }}
+          make build
+
+      - name: Docker push
+        run: |
+          export BRANCH_TAG=${{ env.BRANCH_TAG }}-${{ matrix.variant }}
+          cd $APP_NAME
+          docker-compose push test
+
+      - name: Template test
+        run: |
+          export BRANCH_TAG=${{ env.BRANCH_TAG }}-${{ matrix.variant }}
+          make test_template VARIANT=${{ matrix.variant }}
```

Comment:
IMO, I prefer to move this test to a separate job, so we could separate the jobs to

- Template's unit test
- Variant test
- Production docker image build test

What do you think @hoangmirs

Reply:
With the current implementation, we still need to create a new app to run the template test, the same as variant test, so I will keep them together for now. We could separate them later when we refactor it